Revamped PatchDescriptorLoad

trenouf opened the pull request:
This commit changes the code to implement image and sampler descriptor
loading with the semantics intended by the separate Builder calls to
1. get the descriptor pointer
2. add an index to a descriptor pointer
3. load the descriptor from the pointer

Before this commit, the code relied on being able to scan back from (3)
to (1) to discover the descriptor set and binding. If we have
non-inlined subroutines in the future, that is no longer possible, as
SPIR-V allows the passing of a descriptor pointer as a call arg.

An immutable sampler is now handled by putting it into a read-only
global variable at (1), and relying on LLVM optimizations to promote it
back to a constant in the common case.

Change-Id: Ic37fcb2858dfbc0a84b065dd7987bd4a2746e0e3

amdrexu commented on llpc/patch/llpcPatchDescriptorLoad.cpp:
```diff
@@ -136,46 +136,270 @@ bool PatchDescriptorLoad::runOnModule(
+        byteSize = DescriptorSizeResource;
+        break;
+    default:
+        llvm_unreachable("");
```
-> llvm_unreachable("Should never be called!")
Just like what you have done for existing codes. If this is too verbose, maybe you can change all of them to llvm_unreachable(""). Anyway, I just want to make codes consistent without varying comments for such assert.

trenouf replied:
The only reason all the other ones have the same message is that they replace an LLPC macro that hard-wired that message. Any new one like this should have either no message or (preferably) a meaningful message. See the llvm coding standards doc and codebase. I have now added a meaningful message here.

trenouf commented on llpc/builder/llpcPipelineState.cpp:
```diff
@@ -829,6 +829,7 @@ void PipelineState::ReadUserDataNodes(
 // node that contains it (or is equal to it).
 std::pair<const ResourceNode*, const ResourceNode*> PipelineState::FindResourceNode(
     ResourceMappingNodeType   nodeType,   // Type of the resource mapping node
+    ResourceMappingNodeType   nodeType2,  // Alternate type to look for
```
Not applicable -- the latest version of the commit does not have this second node type to look for.

trenouf commented:
V2: Fixed BufferCompact and dynamic descriptors in spill table.
    V3: Rebased.
    V4: Fixed indexed descriptors. Addressed review comment.

amdvlk-admin commented:
### Test summary for commit bc34d47579636f3a070a5af0e0e0df81a9e00e2b
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev e642f608a62887d40d1f25509d2951a4a3576985</li>
<li>SPVGEN: dev b30c124e7a042c5251aeb20780a83089355e74a1</li>
<li>XGL: dev 80e5a4b11ad2058097e77746772ddc9ab2118e07</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1ebc431ac299fa294a4e730432520d37f326843c</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 32991/33276 (99.1%)</li>
<li><details><summary>Failed: 285/33276 (0.9%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.pipeline.push_constant.graphics_pipeline.count_5_shaders_vert_tess_geom_frag
Stack trace: Image mismatch

FAILURE: dEQP-VK.pipeline.push_constant.graphics_pipeline.overlap_4_shaders_vert_tess_frag
Stack trace: Image mismatch

FAILURE: dEQP-VK.pipeline.push_constant.graphics_pipeline.overlap_5_shaders_vert_tess_geom_frag
Stack trace: Image mismatch

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.default_value.bool
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.default_value.float
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.default_value.double
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.basic.bool
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.basic.uint
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.basic.double
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.builtin.default
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.expression.array_size
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.expression.array_size_length
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.vec2
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.vec3
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.vec4
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.bvec3
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.ivec3
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.ivec4
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.uvec3
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.matrix.mat2
Stack trace: Values did not match
...
```
</p></details></li>


</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 32794/32959 (99.5%)</li>
<li><details><summary>Failed: 165/32959 (0.5%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.pipeline.push_constant.graphics_pipeline.count_5_shaders_vert_tess_geom_frag
Stack trace: Image mismatch

FAILURE: dEQP-VK.pipeline.push_constant.graphics_pipeline.overlap_4_shaders_vert_tess_frag
Stack trace: Image mismatch

FAILURE: dEQP-VK.pipeline.push_constant.graphics_pipeline.overlap_5_shaders_vert_tess_geom_frag
Stack trace: Image mismatch

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.default_value.bool
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.default_value.float
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.default_value.double
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.basic.bool
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.basic.uint
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.basic.double
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.builtin.default
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.expression.array_size
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.expression.array_size_length
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.vec2
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.vec3
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.vec4
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.bvec3
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.ivec3
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.ivec4
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.uvec3
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.matrix.mat2
Stack trace: Values did not match
...
```
</p></details></li>


</ul></details>

<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 33003/33286 (99.1%)</li>
<li><details><summary>Failed: 283/33286 (0.9%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.pipeline.push_constant.graphics_pipeline.count_5_shaders_vert_tess_geom_frag
Stack trace: Image mismatch

FAILURE: dEQP-VK.pipeline.push_constant.graphics_pipeline.overlap_4_shaders_vert_tess_frag
Stack trace: Image mismatch

FAILURE: dEQP-VK.pipeline.push_constant.graphics_pipeline.overlap_5_shaders_vert_tess_geom_frag
Stack trace: Image mismatch

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.default_value.bool
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.default_value.float
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.default_value.double
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.basic.bool
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.basic.uint
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.basic.double
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.builtin.default
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.expression.array_size
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.expression.array_size_length
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.vec2
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.vec3
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.vec4
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.bvec3
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.ivec3
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.ivec4
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.vector.uvec3
Stack trace: Values did not match

FAILURE: dEQP-VK.pipeline.spec_constant.graphics.tess_eval.composite.matrix.mat2
Stack trace: Values did not match
...
```
</p></details></li>


</ul></details>


</ul></details>

nhaehnle commented:
I haven't gone over it with a fine comb, but this seems reasonable to me.

trenouf commented:
V5: Fixed ring descriptors.
    V6: Fixed image from combined texture with immutable sampler, including
        array.
    V7: Fixed lit tests.
    V8: Fixed userDataLimit with spill.
    V9: Fixed inline buffer.

trenouf commented:
Ping for review.

trenouf commented:
This looks spurious. I got a bunch of VK_ERROR_INCOMPATIBLE_DRIVER failures only on one agent.

trenouf commented:
retest this please
